Start Keploy server from Typescript-SDK

frey0-0 opened the pull request:
Fixes [#352](https://github.com/keploy/keploy/issues/352)
If the keploy mode is record and test then it will check if keploy is running on port 6789 and if it isn't then it will start an asynchronously process to start the keploy server and the server won't keeping running, it will stop the server as soon as the main app server stops

frey0-0 committed:
fix(middleware): run keploy binary in middleware

Signed-off-by: Tanmay Bajaj <tanmaybajaj567@gmail.com>

re-Tick commented:
Hi @Frey0-0, we should start the Keploy server from Keploy class constructor in src/keploy.ts because middleware is called on API call only. But we should check for Keploy binary during user API server setup.

Frey0-0 commented:
Yes that makes sense @re-Tick cause it may be possible the binary is not even present. I will look into this and mention if I have any issues here only. Thanks for the review 😄

frey0-0 committed:
feat(constructor): spawn keploy server if mode is test or record

Signed-off-by: Tanmay Bajaj <tanmaybajaj567@gmail.com>

Frey0-0 commented:
@re-Tick I have made the changes now we check if the `keploy` binary is present or not and next if it is running on port `6789` and if it is present and not running then it will spawn a process to start the server

re-Tick commented:
Hi @Frey0-0, I have tried these changes in my local but the keploy binary do not run from the spawn method call.

frey0-0 committed:
fix(logs): remove unwanted logs

Signed-off-by: Tanmay Bajaj <tanmaybajaj567@gmail.com>

Frey0-0 commented:
the spawn method call will run the binary as long as the node server is running and it will be terminated if the node server is stopped because the parent process has stopped. Please have a look at this video below. 

This is on my system which has the keploy binary installed
[Screen Recording](https://user-images.githubusercontent.com/94757729/224575122-a11247b6-6a81-4829-91fa-7ab246afb94d.webm)

This is a system which doesn't have the keploy binary
![Screenshot from 2023-03-13 02-52-04](https://user-images.githubusercontent.com/94757729/224575199-2648fd1f-4605-416d-b1d6-5df1d588b40b.png)

The keploy server stops if we stop the node server and can be started from the command again
[Screen Recording ](https://user-images.githubusercontent.com/94757729/224575543-f0e3e4bc-350f-4011-a046-96137e5579dd.webm)

re-Tick commented:
I have found an interesting issue due to which Keploy binary is trying to run on the `same port` as the sample API server. This is because Keploy runs on port defined by `PORT` env variable or 6789 (default). 

So, we can solve this by changing the env variable name in [keploy](https://github.com/keploy/keploy/blob/ee0e9c33ebeedaf213713e03205343146e21449d/server/server.go#L66) to "KPORT" 

You can repklicate this by: 
1. Use node-fetch sample of samples/typescript.
2. Add `export PORT=8080` in the [.env](https://github.com/keploy/samples-typescript/blob/main/node-fetch/.env) file of node-fetch.
3. Now,  Change the env varibale name from `KEPLOY_APP_PORT` to `PORT` in [this](https://github.com/keploy/samples-typescript/blob/7a26737be3aa0f1f65a898fc0710095f1e34a3e1/node-fetch/server.js#L31) line of app.listen
4. Try recording some testcases. 
Error will be thrown `failed to post testcase to keploy server.`

Frey0-0 commented:
Oh yes! I faced the same error and we should change the name of the Keploy Port. Should I add an issue and then raise a PR for this or directly raise a PR for this describing the issue for the variable name change?

Frey0-0 commented:
I have created an issue and also linked a PR for it. Please review it :) and are there any more changes that need to be done on this PR?

re-Tick committed:
feat(mongoose): addshooks for stubbing mongoose methods (#55)

* fix: stores correct http response for unique test-ids

Signed-off-by: re-Tick <jain.ritik.1001@gmail.com>

* feat(mongoose): adds hooks for mocking/stubbing mongoose methods

Signed-off-by: re-Tick <jain.ritik.1001@gmail.com>

* docs: updates the docs for mongoose and refactors

Signed-off-by: re-Tick <jain.ritik.1001@gmail.com>

* docs: refactors readme for e2e and mocks-lib docs

Signed-off-by: re-Tick <jain.ritik.1001@gmail.com>

* docs: adds require stmt in docs code snippet

Signed-off-by: re-Tick <jain.ritik.1001@gmail.com>

---------

Signed-off-by: re-Tick <jain.ritik.1001@gmail.com>
Signed-off-by: Tanmay Bajaj <tanmaybajaj567@gmail.com>

re-Tick committed:
docs: updates description for router/dependencies

Signed-off-by: Tanmay Bajaj <tanmaybajaj567@gmail.com>

re-Tick committed:
fix(express): checks for the valid resp in send (#65)

Signed-off-by: re-Tick <jain.ritik.1001@gmail.com>
Signed-off-by: Tanmay Bajaj <tanmaybajaj567@gmail.com>

re-Tick commented on src/keploy.ts:
```diff
@@ -76,6 +77,25 @@ export default class Keploy {
+        } else {
+          exec("lsof -i:6789", (error, stdout, stderr) => {
+            if (!stdout) {
+              spawn("keploy");
```
Here, run the Keploy binary with `KEPLOY_PORT=6789` env variable.

Frey0-0 replied:
So I should use the command `export KEPLOY_PORT=6789` before spawning the keploy binary right?

re-Tick replied:
Yes, this export cmd can be run with keploy using ` && ` in the exec method

Frey0-0 replied:
I have done the changes please have a look :sparkles:

re-Tick replied:
LGTM. Please rebase your commit messages for the github action to pass

Frey0-0 replied:
I have made the changes to commit messages. I didn't know there was a specific way to write the commit messages. Thank you for being so patient :sweat_smile:

re-Tick committed:
fix(src/utils): adds ctx parameter in processDep (#69)

the executionContext was getting undefined for mocking dependency call outputs

Signed-off-by: re-Tick <jain.ritik.1001@gmail.com>
Signed-off-by: Tanmay Bajaj <tanmaybajaj567@gmail.com>

frey0-0 committed:
fix: export KEPLOY_PORT while starting keploy server

Signed-off-by: Tanmay Bajaj <tanmaybajaj567@gmail.com>

frey0-0 committed:
Merge branch 'main' into start-sdk-keploy

re-Tick commented:
LGTM. Please add the link in GSOC task list
